### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by artemsen**
Set of scripts to simplify access to OpenBMC setup.

Signed-off-by: Artem Senichev <a.senichev@yadro.com>

### CODE REVIEW

File: admin/led [hunk 1]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash -eu
```
Comment (nest1ing) [inline admin/led:1]: It is really required bash?

File: admin/user [hunk 2]
```diff
@@ -0,0 +1,152 @@
+#!/bin/bash -eu
+
+#
+# User manager.
+#
+
```
Comment (nest1ing) [inline admin/user:6]: This script repeats a functionality of xyz.openbmc_project.User.Manager service.
I think it would be better to use D-Bus interfaces instead direct calls of the system utilities.

File: install.sh [hunk 3]
```diff
@@ -0,0 +1,111 @@
+#!/bin/sh -eu
+
+#
+# Installer.
+#
+
+INSTALL_ROOT=""
+MACHINE="nicole"
```
Comment (nest1ing) [inline install.sh:8]: for compatibility with other platforms this value can be read from `/etc/os-release`

**Review by nest1ing — CHANGES_REQUESTED** [review]

File: profile.in [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+# replace built-in help command
+alias help="/usr/bin/help"
+
+# add sbin to path
+PATH="${PATH}:/sbin:/usr/sbin"
+
+# greeting
+cat << EOF
+Welcome to BMC for Nicole server!
```
Comment (nest1ing) [inline profile.in]: And here, the machine name also should be read from `/etc/os-release`

File: admin/led (hunk 1 above)
Comment (artemsen) [reply admin/led:1]: Yes, there is a bashims, such as [[.
All scripts use bash (except installer).

File: install.sh (hunk 3 above)
Comment (artemsen) [reply install.sh:8]: This script is only for installation purpose, it is not used in runtime.

File: admin/user (hunk 2 above)
Comment (artemsen) [reply admin/user:6]: Yes, I have thought about that too.
But the service has complex interface, it is hard to communicate with them via bash. The second thing - it doesn't cover all our requirements.
I would suggest using this script now, later we can create a standalone module to manage user accounts via console.

File: admin/led (hunk 1 above)
Comment (nest1ing) [reply admin/led:1]: I'm not sure that it is strongly required, but ok, let it be.
we can rewrite it later ;)

File: install.sh [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+echo "Install Phosphor CLI environment to ${INSTALL_ROOT}"
+
+THIS_DIR="$(realpath $(dirname $0))"
+
+install_exec_dir ${THIS_DIR}/admin ${GROUP_ADMIN}
+install_exec_dir ${THIS_DIR}/operator ${GROUP_OPERATOR}
+install_exec_dir ${THIS_DIR}/user ${GROUP_USER}
+
+echo "Install help"
+install_exec_file ${THIS_DIR}/help.in ${INSTALL_ROOT}/usr/bin/help
```
Comment (nest1ing) [inline install.sh]: Just to note: the help conflicts with the bash built-in command. So, it can be called only by full path.

Comment (artemsen) [reply install.sh]: That's why I have added alias to the profile file.

File: admin/user (hunk 2 above)
Comment (nest1ing) [reply admin/user:6]: I fear that the user created by system tools will not be managed by service until it restarted.

Comment (artemsen) [reply admin/user:6]: It's true, D-Bus service doesn't know about new accounts until restart.

Comment (artemsen) [reply admin/user:6]: But it is not a problem - we can restart the service from the script.

Comment (artemsen) [reply admin/user:6]: Done. Works as expected with service restart.

**Review by nest1ing — APPROVED** [review]

File: user/sensors [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+      echo "Invalid argument: $1" >&2
+      exit 1;;
+  esac
+  shift
+done
+
+# elevate self privileges
+[[ ${EUID} -eq 0 ]] || exec sudo $(realpath ${BASH_SOURCE[0]})
+
+exec lssensors
```
Comment (nest1ing) [inline user/sensors]: this command can show sensors of specified type only.
It would be nice if you add this possibility here too.

Comment (artemsen) [reply user/sensors]: Added as optional parameter.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# Phosphor Command-line interface for mere mortals
+Phosphor CLI is a Bash-based shell. This shell is used by non-system users
```
Comment (AlexanderAmelkin) [inline README.md]: I guess it's `YADRO OpenBMC Command Line Interface` and `YADRO OpenBMC CLI`.
Let's also not be this pretensious and just say 'end users' instead of 'mere mortals'.

File: README.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+to manage the OpenBMC system.
+This project contains scripts that simplify access to system setup.
+
+## Access permission
+Different users may have different access rights. Some of them need access
+to system resources that belong to root.
+
+Current implementation founded on Linux groups and `sudo` usage.
```
Comment (AlexanderAmelkin) [inline README.md]: `is based` instead of `founded`

**Review by nest1ing — APPROVED** [review]

File: README.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+OpenBMC has three predefined groups:
+- `priv-admin` (administrators);
+- `priv-operator` (operators);
+- `priv-user` (others).
+
+Each group has its own permissions. The installer (`./install.sh`) creates a
+sudo configuration for each script - links between executable modules and
+Linux groups. This configuration allows group members to run scripts with
```
Comment (AlexanderAmelkin) [inline README.md]: "allows **for** group member to run"
https://english.stackexchange.com/questions/30069/whats-the-difference-between-allow-and-allow-for

File: admin/config [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+#
+
+# prevent modification of executable path
+PATH="/bin:/sbin:/usr/bin:/usr/sbin"
+
+# elevate self privileges
+[[ ${EUID} -eq 0 ]] || exec sudo $(realpath ${BASH_SOURCE[0]}) $@
+
+# list of files to save/restore
+WHITELIST="/var/run/initramfs/whitelist"
```
Comment (AlexanderAmelkin) [inline admin/config]: I'm not sure that this is a good list to pack. It includes things like `requestedHostTransition`, for instance. I don't think those need to be backed up as part of configuration.

File: admin/config [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+# parse command line options
+ACTION="help"
+[[ $# -eq 0 ]] || ACTION="$1"
```
Comment (AlexanderAmelkin) [inline admin/config]: Please make it `-ge 1` or `-ne 0`, and `&&`, or better yet, make it a 3-line `if;then;fi`.

File: admin/config [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    echo "Backup or restore OpenBMC configuration."
+    echo "Use: ${0##*/} [OPTION...]"
+    echo "  backup FILE   Save configuration to file"
+    echo "  restore FILE  Restore configuration from file"
+    echo "  reset         Reset configuration to manufacturing default"
+    echo "  -h, --help    Print this help and exit"
+    ;;
+
+  backup )
+    [[ $# -eq 2 ]] || { echo "Invalid arguments" >&2; exit 1; }
```
Comment (AlexanderAmelkin) [inline admin/config]: Again, please don't use inverse logic in conditionals unless absolutely neccessary.
Why not `[[ $# -ne 2 ]]` ?
Also, it would be nice to just `shift` all the arguments once we've got the `ACTION` and then use `$1` for the first argument of the action. I would also fetch arguments into named variables and use them instead of positionals. That adds clarity to code.

File: admin/led [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+      else
+        echo "Invalid argument, action already defined" >&2
+        exit 1
+      fi;;
+    * )
+      echo "Invalid argument: $1" >&2
+      exit 1;;
+  esac
+  shift
+done
```
Comment (AlexanderAmelkin) [inline admin/led]: So, what was the purpose of the `while` loop? As far as I can see, we now can use `led on status off`, yet the action will be `on` in the end. I fail to understand for what reason we could possibly need that.

File: admin/led [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+# elevate self privileges
+[[ ${EUID} -eq 0 ]] || exec sudo $(realpath ${BASH_SOURCE[0]}) ${ACTION}
```
Comment (AlexanderAmelkin) [inline admin/led]: Inverse logic again

File: admin/led [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+# D-Bus object descriptions to manage the LED
+DBUS_SERVICE="xyz.openbmc_project.LED.GroupManager"
+DBUS_OBJECT="/xyz/openbmc_project/led/groups/enclosure_identify"
+DBUS_INTERFACE="xyz.openbmc_project.Led.Group"
+DBUS_PROPERTY="Asserted"
```
Comment (AlexanderAmelkin) [inline admin/led]: Move all configuration to the top?

File: admin/config [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+#
+# BMC config backup/restore.
+#
+
+# prevent modification of executable path
+PATH="/bin:/sbin:/usr/bin:/usr/sbin"
+
+# elevate self privileges
+[[ ${EUID} -eq 0 ]] || exec sudo $(realpath ${BASH_SOURCE[0]}) $@
```
Comment (AlexanderAmelkin) [inline admin/config]: You have this in all the scripts. Maybe, use some common function?

File: admin/config [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+# list of files to save/restore
+WHITELIST="/var/run/initramfs/whitelist"
+
+# parse command line options
+ACTION="help"
+[[ $# -eq 0 ]] || ACTION="$1"
+case "${ACTION}" in
+  help | -h | --help )
```
Comment (AlexanderAmelkin) [inline admin/config]: A space before the case closing parenthesis? That's a strange coding style.
I'd recommend Google's:
https://google.github.io/styleguide/shellguide.html#s5.5-case-statement

File: admin/led [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+# set state
+STATE=
+case "${ACTION}" in
+  on ) STATE="true";;
+  off ) STATE="false";;
+esac
+if [[ -n "${STATE}" ]]; then
+  busctl set-property ${DBUS_SERVICE} ${DBUS_OBJECT} \
+                      ${DBUS_INTERFACE} ${DBUS_PROPERTY} \
+                      b ${STATE}
```
Comment (AlexanderAmelkin) [inline admin/led:51]: No return code checking? What if `busctl` fails?

File: admin/config [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+    echo "Use: ${0##*/} [OPTION...]"
+    echo "  backup FILE   Save configuration to file"
+    echo "  restore FILE  Restore configuration from file"
+    echo "  reset         Reset configuration to manufacturing default"
+    echo "  -h, --help    Print this help and exit"
+    ;;
+
+  backup )
+    [[ $# -eq 2 ]] || { echo "Invalid arguments" >&2; exit 1; }
+    tar czf "$2" -T ${WHITELIST} &>/dev/null || /bin/true
```
Comment (AlexanderAmelkin) [inline admin/config]: Why mask the `tar`'s return state with a call to `true` and then check the status basing on existence of a file?

File: admin/config [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+      exit 1;
+    fi
+    ;;
+
+  restore )
+    [[ $# -eq 2 ]] || { echo "Invalid arguments" >&2; exit 1; }
+    TMP_DIR="$(mktemp -d)"
+    tar xf "$2" -C ${TMP_DIR}
+    for FILE in $(cat ${WHITELIST}); do
+      [[ -e ${TMP_DIR}${FILE} ]] || continue
```
Comment (AlexanderAmelkin) [inline admin/config]: What's so attractive in inverse logic for you? ;)

File: admin/config [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+    else
+      echo "Error creating backup: $2" >&2
+      exit 1;
+    fi
+    ;;
+
+  restore )
+    [[ $# -eq 2 ]] || { echo "Invalid arguments" >&2; exit 1; }
+    TMP_DIR="$(mktemp -d)"
+    tar xf "$2" -C ${TMP_DIR}
```
Comment (AlexanderAmelkin) [inline admin/config]: If it fails, we just continue?

File: admin/config [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+    for FILE in $(cat ${WHITELIST}); do
+      [[ -e ${TMP_DIR}${FILE} ]] || continue
+      rm -rf ${FILE} &>/dev/null || true
```
Comment (AlexanderAmelkin) [inline admin/config]: I would add checking for the Server's serial number and at least warn if it differs, rather than blindly applying some stray backup archive.
It could also be good to always notify the user of the original system's hostname and serial number before even trying to replace any files.
Then an option like `-y` or `-f` could come in handy to skip any questions.
Then parsing the options could be done with `getopt`. Or maybe not.

File: admin/logs [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+      echo "Invalid argument: $1" >&2
+      exit 1;;
+  esac
+  shift
+done
+
+# elevate self privileges
+[[ ${EUID} -eq 0 ]] || exec sudo $(realpath ${BASH_SOURCE[0]})
+
+exec journalctl --no-hostname --no-pager
```
Comment (AlexanderAmelkin) [inline admin/logs]: I suppose this is just the first version of the script? Asking because this is not what a `logs` command in a CLI is supposed to do. See the proposed CLI design document on the intranet Confluence.

File: admin/nmi.vesnin [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+PATH="/bin:/sbin:/usr/bin:/usr/sbin"
+
+# elevate self privileges
+[[ ${EUID} -eq 0 ]] || exec sudo $(realpath ${BASH_SOURCE[0]}) ${FORCE}
+
+# Color output support
+if [[ -t 1 ]]; then
+  readonly CLR_OK="\e[32m"
+  readonly CLR_ERR="\e[31m"
+  readonly CLR_RST="\e[0m"
```
Comment (AlexanderAmelkin) [inline admin/nmi.vesnin]: This is the only script using colors. Why? Also, there is no guarantee that if stdout is a terminal then it supports ANSI color codes. Some user could be using some Windows-based dumb terminal program to connect to BMC's serial console. They will get garbage then.

File: admin/update [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+# parse command line options
+FILE=
+RESET=
+YES=
+while [[ $# -gt 0 ]]; do
+  case $1 in
+    -h | --help )
+      echo "Update BIOS/BMC firmware."
+      echo "Use: ${0##*/} [OPTION...] FILE"
+      echo "  -r, --reset   Reset all settings to manufacturing default"
```
Comment (AlexanderAmelkin) [inline admin/update]: Please consult the CLI proposal document on intranet Confluence

File: help.in [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+${ADMIN}    && echo -e "led         Chassis LED identify control"
+               echo -e "console     Open host console"
+               echo -e "sensors     Print sensors state"
+${ADMIN}    && [[ -e /usr/bin/nmi ]] && echo -e "nmi         Initiate kdump on the host system"
+${ADMIN}    && echo -e "logs        Print BMC system logs"
+${ADMIN}    && echo -e "sosreport   Create system state report"
+${ADMIN}    && echo -e "esel        Decode eSEL"
+${ADMIN}    && echo -e "user        User managment"
+${ADMIN}    && echo -e "update      Update BIOS/BMC firmware"
+${ADMIN}    && echo -e "config      Save/restore/reset BMC configuration"
```
Comment (AlexanderAmelkin) [inline help.in]: I don't like the idea of hard-coding the commands in the `help` script.
There must be a generic method of listing all the installed commands, along with their privileges, and each one of them must be able to provide a short description. To make things faster, their output could be gathered once (maybe even during firmware build by the `install` script) and stored in a file that the `help` script would display.

For example, the generated help file could look like:
```
led "Chassis LED identify control"
...
config "Save/restore/reset BMC configuration"
```
and then the help script could list all the installed scripts along with their groups (i.e. `priv-admin`, `priv-operator`, `priv-user`) and take the descriptions from the the help file filtering it by the current user's privilege.

File: install.sh (hunk 3 above)
Comment (AlexanderAmelkin) [inline install.sh:8]: If this is a build time default (when `-m` is not supplied), please add a comment explaining that.

File: operator/power [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+  shift
+done
+
+# elevate self privileges
+[[ ${EUID} -eq 0 ]] || exec sudo $(realpath ${BASH_SOURCE[0]}) ${FORCE} ${ACTION}
+
+# execute command
+case "${ACTION}" in
+  on )
+    obmcutil poweron;;
```
Comment (AlexanderAmelkin) [inline operator/power]: JFYI, there is no obmcutil on VEGMAN. `ipmitool` is used there.

File: profile.in [hunk 28]
```diff
@@ -0,0 +1,13 @@
+# Default profile
+
+# replace built-in help command
+alias help="/usr/bin/help"
```
Comment (AlexanderAmelkin) [inline profile.in:4]: need a way to call the built-in help too.

**Review by AlexanderAmelkin — CHANGES_REQUESTED** [review]

File: README.md (hunk 7 above)
Comment (artemsen) [reply README.md]: Fixed

File: README.md (hunk 9 above)
Comment (artemsen) [reply README.md]: Interesting. Thanks!

File: admin/config (hunk 10 above)
Comment (artemsen) [reply admin/config]: There is a lot of problems actually. We need more complex solution, but it is hard to implement such logic in bash.
I doubt this script is really useful, but it's better than nothing. Some kind of experimental feature.

File: admin/config (hunk 11 above)
Comment (artemsen) [reply admin/config]: Fixed

File: README.md (hunk 8 above)
Comment (artemsen) [reply README.md]: Fixed

File: admin/config (hunk 17 above)
Comment (artemsen) [reply admin/config]: Fixed.
But I don't like this style, Google guide doesn't cover handling of multiple values, their style looks nice only for single values in case.

File: admin/led (hunk 18 above)
Comment (artemsen) [reply admin/led:51]: If it fails, busctl will print error message and script execution will be terminated (see [set -e](https://pubs.opengroup.org/onlinepubs/9699919799.2018edition/utilities/V3_chap02.html#set)).

File: admin/config (hunk 19 above)
Comment (artemsen) [reply admin/config]: Tar returns an error code because it can't find all the files from whitelist, some of them may still be missing.

File: admin/config (hunk 16 above)
Comment (artemsen) [reply admin/config]: We have only two common string: path setting and exec sudo. It's not optimal to create common library for two lines of code.
Anyway, we should avoid sudo usage. Sudo is not intended to be used for our purposes, it's a temporary hack.

File: admin/led (hunk 15 above)
Comment (artemsen) [reply admin/led]: Fixed

File: admin/config (hunk 20 above)
Comment (artemsen) [reply admin/config]: I like this small format of if-then-fi, it is simple, easy to read and fits in one line.
Inverse logic is used because all these scripts are executed with -e option.

I can replace them with long format if you insist.

File: admin/config (hunk 21 above)
Comment (artemsen) [reply admin/config]: No, shell's `-e` option prevents further execution.

File: admin/led (hunk 14 above)
Comment (artemsen) [reply admin/led]: I've replied [below](https://github.com/YADRO-KNS/obmc-yadro-cli/pull/1#discussion_r412842643).

File: admin/update (hunk 25 above)
Comment (artemsen) [reply admin/update]: Fixed.

File: help.in (hunk 26 above)
Comment (artemsen) [reply help.in]: I don't like current implementation too. Let's put it off until the next iteration, I need time to think about it.

File: install.sh (hunk 3 above)
Comment (artemsen) [reply install.sh:8]: Added clarification to installer's help.

File: admin/config (hunk 10 above)
Comment (nest1ing) [reply admin/config]: `/var/run` is a symlink. Use `/run/initramfs/whitelist` instead.

File: operator/power (hunk 27 above)
Comment (artemsen) [reply operator/power]: Fixed for nicole build. We will port the project for Vegman later.

File: profile.in (hunk 28 above)
Comment (artemsen) [reply profile.in:4]: Added `helpsh` for that.

File: admin/config (hunk 12 above)
Comment (artemsen) [reply admin/config]: Inverse logic caused by `set -e`. Anyway, fixed with long if-then-fi.

File: admin/led (hunk 13 above)
Comment (artemsen) [reply admin/led]: `while` is used to prevent your case:
```
$ led on status off
Invalid argument: status
```

File: admin/config (hunk 22 above)
Comment (artemsen) [reply admin/config]: As I [said](https://github.com/YADRO-KNS/obmc-yadro-cli/pull/1#discussion_r412824747), complex logic requires more flexible language.
About getopt - it's a headache for linux, we have getopt and getopts, but both are ugly, one doesn't support long options, another is bashism and requires `eval`, which is not safe in pair with sudo.

File: admin/logs (hunk 23 above)
Comment (artemsen) [reply admin/logs]: The entire request is the first version. I have been doing it consider Confluence, but it still a concept.

File: admin/nmi.vesnin (hunk 24 above)
Comment (artemsen) [reply admin/nmi.vesnin]: It came from original version of the script from bitbacket.
Removed for now.

File: admin/config (hunk 10 above)
Comment (artemsen) [reply admin/config]: Path replaced.

Comment (AlexanderAmelkin) [reply admin/config]: According to the information from OpenBMC guys, DMTF has approved the need for this feature in RedFish and they propose to have a schema with an implementation-specific blob. What you do here well fits into that schema. What I'm trying to say is that we'd better invent some good solution now and then use it both for CLI and for RedFish.

File: help.in (hunk 26 above)
Comment (AlexanderAmelkin) [reply help.in]: Just don't forget about it.

File: admin/config (hunk 20 above)
Comment (AlexanderAmelkin) [reply admin/config]: No, I'm fine with oneliners, I'm curious why the logic is inverted. I.e. why `if (expected) {} else { reaction }` instead of `if (!expected) {reaction}`, that is:
```
[[ ! -e ${TMP_DIR}${FILE} ]] && continue
```
looks much more natural to me.

File: admin/config (hunk 22 above)
Comment (AlexanderAmelkin) [reply admin/config]: Nothing stops us from using C or C++ here if that is needed. Server's serial number can well be checked in bash though  using `lsinventory` or `fru` from `phosphor-clio`.

File: admin/led (hunk 13 above)
Comment (AlexanderAmelkin) [reply admin/led]: How so? $1 is `on` in that case. You don't need to go any further than $1. If you just remove while/done along with the last shift, you will get exactly the same result.

File: admin/led (hunk 18 above)
Comment (AlexanderAmelkin) [reply admin/led:51]: Missed the `-e`. I see now.

File: admin/config [hunk 29]
```diff
@@ -26,21 +29,31 @@ case "${ACTION}" in
...
-    tar czf "$2" -T ${WHITELIST} &>/dev/null || /bin/true
-    if [[ -f $2 ]]; then
-      echo "Backup created: $2"
+  backup)
+    if [[ $# -ne 2 ]]; then
+      echo "File name is not specified" >&2
+      exit 1
+    fi
+    shift
+    BACKUP_FILE="$1"
```
Comment (AlexanderAmelkin) [inline admin/config]: Well, I meant that this needed to be done before `case`.

File: admin/config (hunk 20 above)
Comment (artemsen) [reply admin/config]: Because of `set -e` - bash terminates execution if the first condition fails.

File: admin/config (hunk 22 above)
Comment (artemsen) [reply admin/config]: The main problem is in services, that uses the non-volatile storage. There are not only config files, but also caches uses in runtime. Some of services may change own files while we trying to save/restore them and it may lead to unpredictable behavior.

**Review by AlexanderAmelkin — CHANGES_REQUESTED** [review]

File: admin/config (hunk 29 above)
Comment (artemsen) [reply admin/config]: In this case we need to pass file name even for `--help`. Right?

File: admin/led (hunk 13 above)
Comment (artemsen) [reply admin/led]: If you remove `while` you can pass any invalid parameters and they will be ignored, so `led on status dddoff abrakadabra` will work. I believe we should prevent that.

File: help.in (hunk 26 above)
Comment (artemsen) [reply help.in]: Fixed. For now, help is constructed dynamically during build.

File: admin/config (hunk 20 above)
Comment (artemsen) [reply admin/config]: Sorry, I forgot about one detail. These conditions have different behavior in script's body and inside functions, so I always use inverse logic to not to be confused.
Fixed.

File: admin/config (hunk 10 above)
Comment (artemsen) [reply admin/config]: I have removed this feature. Current implementation allows user for change root's password and we can not prevent this without using cryptography. So, we need to have a secret key (in EEPROM?) to sign and check backup file.

File: admin/sosreport [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+      ;;
+    *)
+      echo "Invalid argument: $1" >&2
+      exit 1
+      ;;
+  esac
+  shift
+done
+
+dreport
```
Comment (nest1ing) [inline admin/sosreport]: I think that it would be better to use the hostlogger D-Bus interface for creating archive.
It makes also available this archive in the WebUI.
As an example you can look at [our old implementation for Vesnin ](https://b.yadro.com/projects/SRV/repos/klish/browse/vesnin/scripts/logs/create.sh)

Comment (artemsen) [reply admin/sosreport]: Do you mean phosphor-debug-collector D-Bus? It actually does the same - calls dreport file.
But we lose execution ouput of dreport iself. I don't like the idea to hide stdout/stderr for console applications. And what if there is some problems with D-Bus or debug-collector is stopped working?

Comment (nest1ing) [reply admin/sosreport]: > Do you mean phosphor-debug-collector D-Bus? It actually does the same - calls dreport file.

Yes, sure. I've mistyped.

> But we lose execution ouput of dreport iself. I don't like the idea to hide stdout/stderr for console applications. And what if there is some problems with D-Bus or debug-collector is stopped working?

The general goal of my proposal is keeping only one way for this action. 
For showing the output we could call something like `journalctl -u xyz.openbmc_project.Dump.Manager`. D-Bus might be monitored for example through `busctl monitor`. 
And I agree, it does not look as an easy solution. So I don't  insist, I just propose to think about.

File: admin/led [hunk 31]
```diff
@@ -16,10 +18,10 @@ DBUS_PROPERTY="Asserted"
 # parse command line options
 ACTION=
 while [[ $# -gt 0 ]]; do
-  case $1 in
+  case "$1" in
     -h | --help)
       echo "Chassis LED identify control."
-      echo "Use: ${0##*/} [OPTION...] ACTION"
+      echo "Use: $0 [OPTION...] ACTION"
```
Comment (AlexanderAmelkin) [inline admin/led:20]: Ehm... Why did you remove the basename substitution?

File: admin/led (hunk 13 above)
Comment (AlexanderAmelkin) [reply admin/led]: That can be prevented with parameter count checking. The `while` looks like an overkill for that.

File: admin/led [hunk 32]
```diff
@@ -1,9 +1,11 @@
 #!/bin/bash -eu
 
 #
-# Chassis LED identify control.
+# CLI: Chassis LED identify control
 #
 
+# elevate self privileges
+[[ ${EUID} -ne 0 ]] && exec sudo $(realpath ${BASH_SOURCE[0]}) $@
```
Comment (AlexanderAmelkin) [inline admin/led]: I think that privilege elevation after parameter checking was safer. Why did you move it here?

Comment (AlexanderAmelkin) [reply admin/led]: Also, why all the hassle with `realpath` ? Do you use symlinks anywhere? If so, won't running the script by real path with a help argument reveal its real name instead of the symlink that the user have called?

File: admin/sosreport (hunk 30 above)
Comment (AlexanderAmelkin) [reply admin/sosreport]: Artem, please create an issue (in JIRA, not here) to rewrite these tools for working via D-Bus. Let's not do it now to save time and effort, but I support the idea of single control path for all user interfaces.

File: install.sh [hunk 33]
```diff
@@ -69,6 +70,7 @@ install_exec_dir() {
   install --mode 0755 -d ${INSTALL_ROOT}${bin_dir}
   sudo_dir="/etc/sudoers.d"
   install --mode 0750 -d ${INSTALL_ROOT}${sudo_dir}
+  install --mode 0755 -d $(dirname ${INSTALL_ROOT}${HELP_FILE})
 
   sudo_file="${INSTALL_ROOT}${sudo_dir}/$(echo ${group_name} | sed 's/-//g')"
```
Comment (AlexanderAmelkin) [inline install.sh]: maybe call the file `cli.${group_name}` instead of just `${group_name}` ? I think that would be better and would prevent future possible clashes with other software.

File: admin/led (hunk 31 above)
Comment (artemsen) [reply admin/led:20]: We don't need that, I suppose we even shouldn't remove the path. You can have multiple programs with the same name in linux. This basename came from my experiments with chroot, where I have to hide the real executable path and work over some kind of proxy.

**Review by AlexanderAmelkin — CHANGES_REQUESTED** [review]

File: admin/sosreport (hunk 30 above)
Comment (artemsen) [reply admin/sosreport]: As I [said](https://github.com/YADRO-KNS/obmc-yadro-cli/pull/1#discussion_r416575679) I don't like this idea, but I have implemented both options.
BTW, I got this today: `May 06 17:44:01 phosphor-dump-manager[1500]: Error occurred during the sd_event_add_child call`, so I wouldn't recommend to call dreport over D-Bus in production.

File: admin/led (hunk 32 above)
Comment (artemsen) [reply admin/led]: Removed.

File: install.sh (hunk 33 above)
Comment (artemsen) [reply install.sh]: Fixed, but I have replaced the dot with underscore, sudo doesn't like dots.

File: admin/led (hunk 32 above)
Comment (artemsen) [reply admin/led]: > I think that privilege elevation after parameter checking was safer. Why did you move it here?

In this case we have to spend time for checking arguments once more.

File: admin/sosreport (hunk 30 above)
Comment (AlexanderAmelkin) [reply admin/sosreport]: I still think we need a single control path. Errors in that path are a different matter. They must be fixed instead of routed around.

Comment (artemsen) [reply admin/sosreport]: System administrator can't fix such errors on production server.
Do you mean I should remove direct dreport invoke, leave D-Bus call only?

File: admin/sosreport [hunk 34]
```diff
@@ -24,8 +24,54 @@ while [[ $# -gt 0 ]]; do
...
+
+# Create report directly with dreport.
+function report_direct {
+  dreport -v
+  echo
+  echo "Report created: /tmp/obmcdump_NNNNNNNN_NNNNNNNNNN.tar.xz"
+}
+
+# try to gather report over D-Bus, if it fails - call dreport directly
+report_dbus || report_direct
```
Comment (AlexanderAmelkin) [inline admin/sosreport:72]: Now this is good!

File: clicmd [hunk 35]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash -eu
+
+#
+# Base CLI command handler.
+# This script used for setting up the envrinment and elevating priveleges for
```
Comment (AlexanderAmelkin) [inline clicmd]: This script **is** used ... envir**o**nment ...

File: clicmd [hunk 36]
```diff
@@ -0,0 +1,24 @@
...
+# real command to execute
+REAL_CMD="/usr/bin/cli_$(basename "${BASH_SOURCE[0]}")"
+if [[ ! -f "${REAL_CMD}" ]]; then
+  echo "Invalid CLI command: ${BASH_SOURCE[0]}" >&2
+  exit 2
+fi
+
+# elevate privileges
+if [[ ${EUID} -ne 0 ]]; then
+  exec sudo ${BASH_SOURCE[0]} $@
```
Comment (AlexanderAmelkin) [inline clicmd]: This is the third time by now that you use `${BASH_SOURCE[0]}`. Maybe extract it at the start into something like `$THIS` ?

File: install.sh [hunk 37]
```diff
@@ -109,6 +108,7 @@ echo "Install Phosphor CLI environment to ${INSTALL_ROOT}"
 
 THIS_DIR="$(realpath $(dirname $0))"
 
+install -DT --mode 0755 ${THIS_DIR}/clicmd ${INSTALL_ROOT}/usr/bin/clicmd
```
Comment (nest1ing) [inline install.sh]: The permissions of `${real_file}` are `0400` but of the `clicmd` are `0755`.
Lets increase our paranoia and set `0555`. ;)

Comment (artemsen) [reply install.sh]: Let's do it for all executable files in the system! )

**Review by AlexanderAmelkin — CHANGES_REQUESTED** [review]

File: install.sh (hunk 37 above)
Comment (AlexanderAmelkin) [reply install.sh]: Yes, make clicmd 555.

File: clicmd (hunk 36 above)
Comment (artemsen) [reply clicmd]: Done.

File: clicmd (hunk 35 above)
Comment (artemsen) [reply clicmd]: Done.

File: install.sh (hunk 37 above)
Comment (artemsen) [reply install.sh]: Done.

**Review by AlexanderAmelkin — APPROVED** [review]

File: admin/sosreport (hunk 30 above)
Comment (AlexanderAmelkin) [reply admin/sosreport]: No, the way you did is ok. First trying the right way and routing around if it didn't work is fine with me.
